README.rst
wms reverse proxy for 3di scalability
=====================================

prerequisites
-------------

Make sure you have Go 1.x installed. 

Set $GOPATH to anywhere your Go user can write, i.e. EXPORT GOPATH=/home/user/git/go

Fetch some dependencies::

    $ go get github.com/codegangsta/cli
    $ go get github.com/garyburd/redigo/redis


instructions for local deployment
---------------------------------

Then build the executable::

    $ go build -o wmsrp .

To build the executable for Windows (not sure why you would want to, but 
anyway)::

    $ env GOOS=windows GOARCH=amd64 go build -o wmsrp .

To see usage information and the command line options of the generated executable, do::

    $ ./wmsrp -h

Or to run it, do (values are examples)::

    $ /.wmsrp --port=8321 --redis-host=10.0.3.100

You can determine the port yourself, but it has to match the wms url
port in your threedi_server settings. For example::

    THREEDI_WMS_SERVER_URL = 'http://localhost:8321/3di/wms'
    THREEDI_WMS_DATA_URL = 'http://localhost:8321/3di/data'

The --redis-host value should point to your threedi server's redis server.

Add --use-cache for local development setup (because some wms requests do not have
session cookies for some unknown reason).

When you running the wms reverse proxy with the proper settings as stated
above, your wms results will be served through the reverse proxy.

instruction for deployment on staging and production
----------------------------------------------------

Make sure supervisor is installed::

    $ apt-get install supervisor -y

Create directory /usr/local/wmsrevproxy and put the wmsrp binary in it::

    $ mkdir /usr/local/wmsrevprox
    $ cp wmsrp /usr/local/wmsrevprox

Create a /etc/supervisor/conf.d/wmsrevprox.conf file with this content (N.B. change --redis-host to p-3di-red-d1.external-nens.local for production)::

    [program:wms_reverse_proxy]
    command = /usr/local/wmsrevprox/wmsrp -p 6666 --redis-host=s-3di-red-d1.external-nens.local
    process_name = wms_reverse_proxy
    directory = /usr/local/wmsrevprox
    priority = 20
    redirect_stderr = true

Add log rotation to /etc/supervisor/supervisor.conf::

    [supervisord]
    logfile=/var/log/supervisor/supervisord.log
    pidfile=/var/run/supervisord.pid
    childlogdir=/var/log/supervisor
    stdout_logfile_maxbytes = 104857600
    stdout_logfile_backups = 10
    stderr_logfile_maxbytes = 104857600
    stderr_logfile_backups = 10

Make sure the /3di location in the nginx site config file points to the reverse proxy (proxy pass host and port must match wmsrp host and port)::

    location /3di {
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $http_host;
        proxy_redirect off;
        proxy_pass http://127.0.0.1:6666;
        proxy_read_timeout 60;
    }

Reload nginx and restart supervisor::

    $ service nginx reload
    $ service supervisor restart

The wms reverse proxy log file can be found here::

    $ tail -f /var/log/supervisor/wms_reverse_proxy-stdout---supervisor-*.log

required redis variables
------------------------

These redis variables are required::

    # set by the socket server
    - session_to_subgrid_id
    - <subgrid_id>:loaded_model_type

    # set by the machine manager
    - subgrid_id_to_wms_address
    - subgrid_id_to_flow_address


CHANGES.rst
Changelog of wmsrp
==================


0.3 (unreleased)
----------------

- Use ``gb`` for dependency management.


0.2 (2015-11-07)
----------------

- Fetch wms and flow address from the ``subgrid_id_to_wms_address`` and 
  ``subgrid_id_to_flow_address`` respectively.

- Resolve full address (<host or ip>:<port>) to redirect to instead of ip only.

- Use version from ``version.txt`` file to make it possible to release with 
  zest.releaser.


0.1.0 (2015-08-03)
------------------

- Initial release.



wms_reverse_proxy_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestRedisPool(t *testing.T) {
	pool := newRedisPool("localhost:6379", "")
	assert.Equal(t, 3, pool.MaxIdle)
	assert.Equal(t, 240000000000, pool.IdleTimeout) // 240 seconds
}


src/wmsrp/reverse_proxy_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestRedisPool(t *testing.T) {
	pool := newRedisPool("localhost:6379", "")
	assert.Equal(t, 3, pool.MaxIdle)
	// need to cast to int, because the values are different types (int32 and int)
	// see issue https://github.com/stretchr/testify/issues/155
	assert.Equal(t, int(240000000000), int(pool.IdleTimeout)) // 240 seconds
}


wmsrp.go
package main

import (
	"log"
	"net/http"
	"os"
	"strings"

	"github.com/codegangsta/cli"
)

var Version = ""

func main() {
	// also use microseconds in log messages
	log.SetFlags(log.Ldate | log.Ltime | log.Lmicroseconds)

	app := cli.NewApp()
	app.Name = "wmsrp"
	app.Usage = "wms reverse proxy for the 3di scalability architecture"
	app.Version = Version
	app.Authors = []cli.Author{
		cli.Author{Name: "Sander Smits", Email: ""},
	}
	app.Flags = []cli.Flag{
		cli.StringFlag{ // always, required
			Name:  "port, p",
			Value: "5050",
			Usage: "port this reverse proxy serves on",
		},
		cli.StringFlag{
			Name:  "redis-host",
			Value: "127.0.0.1",
			Usage: "redis server host",
		},
		cli.StringFlag{
			Name:  "redis-port",
			Value: "6379",
			Usage: "redis server port",
		},
		cli.StringFlag{
			Name:  "single-server",
			Value: "",
			Usage: "say we want no redis/machine manager stuff; use provided subgrid_id (subgrid:10000) and redirect to localhost",
		},
		cli.BoolFlag{
			Name:  "use-cache",
			Usage: "cache session keys (use for development environments)",
		},
	}
	app.Action = func(c *cli.Context) {
		port := c.String("port")
		redisHost := c.String("redis-host")
		redisPort := c.String("redis-port")
		useCache := c.Bool("use-cache")
		singleServer := c.String("single-server")

		wmsRevProxy := wmsReverseProxy(redisHost, redisPort, useCache, singleServer)

		log.Println("- INFO - starting wms reverse proxy on port", port)
		log.Println("- INFO - using redirect info from redis server on", strings.Join([]string{redisHost, redisPort}, ":"))

		err := http.ListenAndServe(":"+port, wmsRevProxy)
		if err != nil {
			log.Fatal("- ERROR - http.ListenAndServe:", err)
		}
	}
	app.Run(os.Args)
}


reverse_proxy.go
package main

import (
	"log"
	"net/http"
	"net/http/httputil"
	"strings"
	"time"

	"github.com/garyburd/redigo/redis"
)

var (
	sessionKeyCache map[string]string
	pool            *redis.Pool
)

func newRedisPool(server, password string) *redis.Pool {
	return &redis.Pool{
		MaxIdle:     3,
		IdleTimeout: 240 * time.Second,
		Dial: func() (redis.Conn, error) {
			c, err := redis.Dial("tcp", server)
			if err != nil {
				return nil, err
			}
			if password != "" {
				if _, err := c.Do("AUTH", password); err != nil {
					c.Close()
					return nil, err
				}
			}
			return c, err
		},
		TestOnBorrow: func(c redis.Conn, t time.Time) error {
			_, err := c.Do("PING")
			return err
		},
	}
}

func wmsReverseProxy(redisHost string, redisPort string, useCache bool, singleServer string) *httputil.ReverseProxy {
	// normally use redis to find subgrid_id and ip belonging to session
	// if singleServer is defined, subgrid_id = singleServer and redirect to localhost
	var subgridID string

	sessionKeyCache := make(map[string]string)

	redisAddress := strings.Join([]string{redisHost, redisPort}, ":")
	pool = newRedisPool(redisAddress, "")

	director := func(req *http.Request) {
		var startTime = time.Now()
		log.Println("- INFO - start resolving wms address")
		// 1) get the session key
		var sessionKey, addressKey, dataAddress string

		sessionCookie, err := req.Cookie("sessionid")
		remoteAddr := strings.Split(req.RemoteAddr, ":")[0] // skip the port
		if err == nil {
			sessionKey = sessionCookie.Value
			log.Println("- INFO - got session key from request:", sessionKey)
			if useCache == true {
				// put the session key in a cache map by using remoteAddr as key
				sessionKeyCache[remoteAddr] = sessionKey
				log.Println("- INFO - storing session key in cache; remote address:", remoteAddr)
			}
		} else if useCache == true {
			log.Println("- INFO - fetching session key from cache; remote address:", remoteAddr)
			sessionKey = sessionKeyCache[remoteAddr]
			if sessionKey != "" {
				log.Println("- INFO - got session key from cache:", sessionKey)
			}
		}
		if sessionKey == "" {
			log.Println("- ERROR - unable to get a session key")
			req.URL = nil // results in a 500 response, which is what we want here
			return
		}

		// redis connection from pool
		// TODO: how to define conn: *redis.pooledConnection ??
		conn := pool.Get()
		log.Println("- DEBUG - number of active connections in redis pool:", pool.ActiveCount())
		defer conn.Close()

		// 2) get the subgrid_id
		if singleServer == "" {
			subgridID, err = redis.String(conn.Do("HGET", "session_to_subgrid_id", sessionKey))
			if err != nil {
				log.Println("- ERROR - unable to get subgrid id:", err)
				req.URL = nil
				return
			}
			log.Println("- INFO - got subgrid id:", subgridID)
		} else {
			subgridID = singleServer
			log.Println("- INFO - single server mode, subgrid_id:", subgridID)
		}

		// 3) get the loaded_model_type
		loadedModelType, err := redis.String(conn.Do("GET", subgridID+":loaded_model_type"))
		if err != nil {
			// assume loaded_model_type is 3di for backwards compatibility
			log.Println("- WARNING - loaded_model_type not available, fallback to default (3di)")
			loadedModelType = "3di"
		}
		log.Println("- INFO - using loaded model type:", loadedModelType)

		// 4) determine which address key to use based on the loaded model type
		switch loadedModelType {
		case "3di":
			addressKey = "wms_address"
		case "3di-urban":
			addressKey = "flow_address"
		default:
			log.Println("- ERROR - unsupported loaded model type:", loadedModelType)
			req.URL = nil
			return
		}

		// 5) get the data address
		if singleServer == "" {
			dataAddress, err = redis.String(conn.Do("HGET", "subgrid_id_to_"+addressKey, subgridID))
			if err != nil {
				log.Println("- ERROR - unable to get data address:", err)
				req.URL = nil
				return
			}
		} else {
			switch loadedModelType {
			case "3di":
				// 5000 is the default wms port
				dataAddress = "127.0.0.1:5000"
			case "3di-urban":
				// 6000 is the default flow port
				dataAddress = "127.0.0.1:6000"
			default:
				log.Println("- ERROR - unsupported loaded model type:", loadedModelType)
				req.URL = nil
				return
			}
		}

		// redirect this request to the resolved data address
		var endTime = time.Now()
		log.Println("- INFO - resolved data address", dataAddress, "in", endTime.Sub(startTime))
		req.URL.Scheme = "http"
		req.URL.Host = dataAddress
	}
	return &httputil.ReverseProxy{Director: director}
}


src/wmsrp/main.go
package main

import (
	"log"
	"net/http"
	"os"
	"strings"

	"github.com/codegangsta/cli"
)

// placeholder for injecting the real version when compiling by build.sh
var Version = "dev"

func main() {
	// also use microseconds in log messages
	log.SetFlags(log.Ldate | log.Ltime | log.Lmicroseconds)

	app := cli.NewApp()
	app.Name = "wmsrp"
	app.Usage = "wms reverse proxy for the 3di scalability architecture"
	app.Version = Version
	app.Authors = []cli.Author{
		cli.Author{Name: "Sander Smits", Email: ""},
	}
	app.Flags = []cli.Flag{
		cli.StringFlag{ // always, required
			Name:  "port, p",
			Value: "5050",
			Usage: "port this reverse proxy serves on",
		},
		cli.StringFlag{
			Name:  "redis-host",
			Value: "127.0.0.1",
			Usage: "redis server host",
		},
		cli.StringFlag{
			Name:  "redis-port",
			Value: "6379",
			Usage: "redis server port",
		},
		cli.StringFlag{
			Name:  "single-server",
			Value: "",
			Usage: "say we want no redis/machine manager stuff; use provided subgrid_id (subgrid:10000) and redirect to localhost",
		},
		cli.BoolFlag{
			Name:  "use-cache",
			Usage: "cache session keys (use for development environments)",
		},
	}
	app.Action = func(c *cli.Context) {
		port := c.String("port")
		redisHost := c.String("redis-host")
		redisPort := c.String("redis-port")
		useCache := c.Bool("use-cache")
		singleServer := c.String("single-server")

		wmsRevProxy := wmsReverseProxy(redisHost, redisPort, useCache, singleServer)

		log.Println("- INFO - starting wms reverse proxy on port", port)
		log.Println("- INFO - using redirect info from redis server on", strings.Join([]string{redisHost, redisPort}, ":"))

		err := http.ListenAndServe(":"+port, wmsRevProxy)
		if err != nil {
			log.Fatal("- ERROR - http.ListenAndServe:", err)
		}
	}
	app.Run(os.Args)
}


wms_reverse_proxy.go
package main

import (
	"log"
	"net/http"
	"net/http/httputil"
	"os"
	"strings"
	"time"

	"github.com/codegangsta/cli"
	"github.com/garyburd/redigo/redis"
)

var (
	sessionKeyCache map[string]string
	pool            *redis.Pool
)

func newRedisPool(server, password string) *redis.Pool {
	return &redis.Pool{
		MaxIdle:     3,
		IdleTimeout: 240 * time.Second,
		Dial: func() (redis.Conn, error) {
			c, err := redis.Dial("tcp", server)
			if err != nil {
				return nil, err
			}
			if password != "" {
				if _, err := c.Do("AUTH", password); err != nil {
					c.Close()
					return nil, err
				}
			}
			return c, err
		},
		TestOnBorrow: func(c redis.Conn, t time.Time) error {
			_, err := c.Do("PING")
			return err
		},
	}
}

func wmsReverseProxy(redisHost string, redisPort string, wmsPort string, useCache bool) *httputil.ReverseProxy {
	sessionKeyCache := make(map[string]string)

	redisAddress := strings.Join([]string{redisHost, redisPort}, ":")
	pool = newRedisPool(redisAddress, "")

	director := func(req *http.Request) {
		// 1) get the session key
		var sessionKey string
		sessionCookie, err := req.Cookie("sessionid")
		remoteAddr := strings.Split(req.RemoteAddr, ":")[0] // skip the port
		if err == nil {
			sessionKey = sessionCookie.Value
			log.Println("got session key from request:", sessionKey)
			if useCache == true {
				// put the session key in a cache map by using remoteAddr as key
				sessionKeyCache[remoteAddr] = sessionKey
				log.Println("storing session key in cache; remote address:", remoteAddr)
			}
		} else if useCache == true {
			log.Println("fetching session key from cache; remote address: ", remoteAddr)
			sessionKey = sessionKeyCache[remoteAddr]
			if sessionKey != "" {
				log.Println("got session key from cache:", sessionKey)
			}
		}
		if sessionKey == "" {
			log.Println("unable to get a session key")
			req.URL = nil // results in a 500 response, which is what we want here
			return
		}

		// redis connection from pool
		conn := pool.Get()
		log.Println("number of active connections in redis pool:", pool.ActiveCount())
		defer conn.Close()

		// 2) get the subgrid_id
		subgridID, err := redis.String(conn.Do("HGET", "session_to_subgrid_id", sessionKey))
		if err != nil {
			log.Println("unable to get subgrid id:", err)
			req.URL = nil
			return
		}
		log.Println("got subgrid id:", subgridID)

		// 3) get the wms server ip
		wmsIP, err := redis.String(conn.Do("HGET", "subgrid_id_to_ip", subgridID))
		if err != nil {
			log.Println("unable to get wms ip:", err)
			req.URL = nil
			return
		}

		// use the full wms address to redirect this request to
		wmsAddress := strings.Join([]string{wmsIP, wmsPort}, ":")
		log.Println("got wms address:", wmsAddress)
		req.URL.Scheme = "http"
		req.URL.Host = wmsAddress
	}
	return &httputil.ReverseProxy{Director: director}
}

func main() {
	app := cli.NewApp()
	app.Name = "wmsrp"
	app.Usage = "wms reverse proxy server for the 3di scalability architecture"
	app.Version = "0.0.1"
	app.Authors = []cli.Author{
		cli.Author{Name: "Sander Smits", Email: ""},
	}
	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:  "port, p",
			Value: "5050",
			Usage: "port this reverse proxy serves on",
		},
		cli.StringFlag{
			Name:  "redis-host",
			Value: "127.0.0.1",
			Usage: "redis server host",
		},
		cli.StringFlag{
			Name:  "redis-port",
			Value: "6379",
			Usage: "redis server port",
		},
		cli.StringFlag{
			Name:  "wms-port",
			Value: "5000",
			Usage: "wms server port",
		},
		cli.BoolFlag{
			Name:  "use-cache",
			Usage: "cache session keys (use for development environments)",
		},
	}
	app.Action = func(c *cli.Context) {
		port := c.String("port")
		redisHost := c.String("redis-host")
		redisPort := c.String("redis-port")
		wmsPort := c.String("wms-port")
		useCache := c.Bool("use-cache")

		wmsRevProxy := wmsReverseProxy(redisHost, redisPort, wmsPort, useCache)

		log.Println("starting wms reverse proxy on port", port)
		log.Println("using redirect info from redis server on", strings.Join([]string{redisHost, redisPort}, ":"))
		log.Println("redirecting to wms servers on port", wmsPort)

		err := http.ListenAndServe(":"+port, wmsRevProxy)
		if err != nil {
			log.Fatal("http.ListenAndServe:", err)
		}
	}
	app.Run(os.Args)
}
